### PR TITLE
consoleproxy: Enable console for vms in Stopping/Migrating state

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

Currently vm console is not enabled if vm is in stopping or migrating.
Enabling vm console will help us 

Fixes: #3631

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by ustcweizhou** [comment]
@andrijapanicsb this fixes your issue #3631

**Comment by svenvogel** [comment]
@ustcweizhou nice enhancement!

**Comment by andrijapanicsb** [comment]
_I need to improve my Outlook filter (to see some of those almost 10.000 emails)_

Thanks @ustcweizhou
@blueorangutan package

**Comment by andrijapanicsb** [comment]
Tested, looks fine, thx @ustcweizhou 

LGTM

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by svenvogel** [comment]
@weizhouapache sounds intresting. not tested but i hope it will go into the master. we encounter the same problem.

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by ustcweizhou** [comment]
@rhtyd @andrijapanicsb can you kick off a test again ?

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
need one more LGTM from code POV, otherwise tested manually fine and tests are passing.

File: server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManagerImpl.java [hunk 1]
```diff
@@ -401,9 +401,10 @@ public ConsoleProxyVO doAssignProxy(long dataCenterId, long vmId) {
             return null;
         }
 
-        if (vm != null && vm.getState() != State.Starting && vm.getState() != State.Running) {
+        if (vm != null && vm.getState() != State.Starting && vm.getState() != State.Running
+                && vm.getState() != State.Stopping && vm.getState() != State.Migrating) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManagerImpl.java]: :) a lot of getState()s there

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java [hunk 2]
```diff
@@ -2392,7 +2392,12 @@ public String getConsoleAccessUrlRoot(final long vmId) {
             s_logger.trace("Trying to retrieve VNC port from agent about VM " + vm.getHostName());
         }
 
-        final GetVncPortAnswer answer = (GetVncPortAnswer)_agentMgr.easySend(vm.getHostId(), new GetVncPortCommand(vm.getId(), vm.getInstanceName()));
+        GetVncPortAnswer answer = null;
+        if (vm.getState() == State.Migrating && vm.getLastHostId() != null) {
+            answer = (GetVncPortAnswer)_agentMgr.easySend(vm.getLastHostId(), new GetVncPortCommand(vm.getId(), vm.getInstanceName()));
+        } else {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/ManagementServerImpl.java]: So we are making the decision we are opening a console to the old VM. looks good but may have some problems. I suspect you tested @ustcweizhou .
problems I see are:
- this is a second migrate and the old host is no longer valid.
- the migration is already successful but not administered yet.

**Comment by DaanHoogland** [comment]
travis job 5 ran out of log space. this has happened more often recently. no issue concerning this PR afaiac

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java (hunk 2 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/server/ManagementServerImpl.java]: @DaanHoogland indeed there are some problems with vm console, for example
(1) vm is Starting state, but is not running on any server  (eg applying dhcp/dns entries in virtual routers)
(2) vm is Stopping state, but is stopped on server
(3) vm is Migrating state, when we open the console of vm (on source host), and migration is done, the console is invalid.
However, I think all these problems are not real issues. When vm state is updated, refresh the web page and a new console will work.

File: server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManagerImpl.java]: @DaanHoogland not very much , only 4 :-) 
will pay attention in the future.

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManagerImpl.java]: NP, code can be such a joy.

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/server/ManagementServerImpl.java]: I agree, @weizhouapache these are not big issues. I just wanted to have mentioned them and am not 👎ing this!

**Pull request merged**